### ISSUE TITLE
Consider a hot reload feature switch for linking

### ISSUE THREAD

**Issue by stephentoub** [dotnet/runtime#51159]
## Background and Motivation

We are going to have a variety of places across the core libraries that use MetadataUpdateHandlerAttribute to react to hot reload updates. Some of these handlers will have zero impact unless used (other than their presence in the library binaries) while others will have additional code on "normal" execution code paths, e.g. to track things that later need to be cleared. We should look at either using an existing feature switch (e.g. Debugging.IsSupported) or a new feature switch (e.g. AssemblyExtensions.IsApplyUpdateAvailable) that we tie all of these to and use to both link away and disable any expense when not required. We should aim to use the same switch across the whole stack (currently ASP.NET is using its own property).

## Proposed API
```C#
namespace System.Reflection.Metadata
{
    public static class AssemblyExtensions
    {
        /// <summary>
        /// Returns true if the apply assembly update is enabled and available.
        /// </summary>
        public static bool IsApplyUpdateAvailable { get; }
    }
}
```

The .NET Core runtime will basically implement like this (and will cache the value in a static):

```C#
    IsApplyUpdateAvailable => Debugger.IsAttached || Environment.GetEnvironmentVariable("DOTNET_MODIFIABLE_ASSEMBLIES") != "" || Environment.GetEnvironmentVariable("COMPlus_ForceEnc") == "1";
```

The feature switch support in the runtime build files include this property so the conditional apply update code is elided on release builds.

## Usage Examples

```C#
    if (AssemblyExtensions.IsApplyUpdateAvailable)
    {
        // Hot reload/ENC specific initialization or bookkeeping code
    }
```

The hot reload agent can use this to ensure that it can actually apply updates to the application. Third party libraries can use it to run some apply update specific initialization or bookkeeping code.

## Risks

Low. This API is used only in very limited cases and by very few developers (mostly internal).

**Comment by pranavkm** [comment]
If I understand this correctly, would we be teaching the trimmer to remove all code paths inside a `HotReload.IsSupported` switch?

**Comment by SamMonoRT** [comment]
@lambdageek fyi

**Comment by eerhardt** [comment]
>  would we be teaching the trimmer to remove all code paths inside a HotReload.IsSupported switch?

If we added this switch, the trimmer would only remove code paths inside the `HotReload.IsSupported` switch when the switch was set to `false`.

There are other scenarios where the trimmer runs, but you still want to debug and hot-reload your app. One that comes to mind is Xamarin apps - See https://github.com/xamarin/xamarin-android/blob/main/Documentation/guides/OneDotNet.md#dotnet-build--publish.

**Comment by marek-safar** [comment]
Is the logic fully detached from Debugging.IsSupported feature switch? In other words would settings like Debugging.IsSupported = false, HotReload.IsSupported = true work?

**Comment by pranavkm** [comment]
Do we still plan on doing something targeted here? ASP.NET Core is currently using `Debugging.IsSupported` for Blazor and I was planning to proliferate it in more places.

**Comment by lambdageek** [comment]
I think the need for a separate feature flag is a bit more acute with the ongoing [componentization of Mono](https://github.com/dotnet/runtime/issues/50862), workloads can decide whether to bundle hot reload support in some configurations or builds of the user's app.

So framework or library authors really need a separate feature switch if they want to affect runtime behavior. (I think we're all aware of the open issue for that, but just in case it's dotnet/runtime#51159)

**Comment by terrajobst** [comment]
* At this point it seems better to make a dedicated type
    - Let's move the already approved `AssemblyExtensions.ApplyUpdate()`
* Let's rename the property to `IsSupported`
* We may want to leave the old APIs until preview 7 so that we keep infrastructure working

```C#
namespace System.Reflection.Metadata
{
    public static class MetadataUpdater
    {
        public static void ApplyUpdate(Assembly assembly, ReadOnlySpan<byte> metadataDelta, ReadOnlySpan<byte> ilDelta, ReadOnlySpan<byte> pdbDelta);
        public static bool IsSupported { get; }
    }
}
```

**Comment by mikem8361** [comment]
We would need to add one more related API: GetApplyUpdateCapabilities() and probably rename it to GetCapabilities(). It is currently internal so the tests and hot reload agent/Roslyn use reflection but maybe we should make it public.

**Comment by stephentoub** [comment]
> We would need to add one more related API: GetApplyUpdateCapabilities() and probably rename it to GetCapabilities()

Yup, we discussed that, it was just left out of the summary.  We landed on GetCapabilities().  We didn't discuss changing it to be public.

We also talked about how we may want to stage this by first adding the new class and changing the existing one to just delegate to it but not yet remove it, wait until the rest of the components have been updated to use the new type, and only then delete the old one (but hopefully that cycle can happen relatively quickly, as we'd want to delete the old one soon-ish).
